---
title: "Exercise 1"
author:
        - "Shaqed Carasso" 
        - "ID: 201281805"
date: "10/28/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Computations in R

1. Sum 2 and 3 using the + operator. 
```{r }
2+3
```
2. Take the square root of 36, use sqrt().
```{r }
sqrt(36)
```
3. Take the log10 of 1000, use function log10(). 
```{r }
log10(1000)
```
4. Take the log2 of 32, use function log2(). 
```{r }
log2(32)
```
5. Assign the sum of 2,3 and 4 to variable x. 
```{r }
x = 2+3+4
x
```
6. Find the absolute value of the expression 5 - 145 using the abs() function.
```{r }
abs(5-145)
```
7. Calculate the square root of 625, divide it by 5, and assign it to variable x.
```{r }
x = sqrt(625)/5
x
```
8. Multiply the value you get from previous exercise by 10000, assign it to variable x 
```{r }
x = 10000*x
x
```


## Data structures in R

10. Make a vector of 1,2,3,5 and 10 using c(), and assign it to the vec variable. 
```{r }
vec = c(1,2,3,5)
vec
```
11. Check the length of your vector with length(). Ex: length(vec1) should return 3.
```{r }
length(vec)
```
12. Make a vector of all numbers between 2 and 15. 
```{r }
c(2:15)
```
13. Make a vector of 4s repeated 10 times using the rep() function. 
```{r }
rep(x = 4, times = 10)
```
14. Make a logical vector with TRUE, FALSE values of length 4, use c().
```{r }
c(T,T,T,F)
```
15. Make a character vector of the gene names PAX6,ZIC2,OCT4 and SOX2.
```{r }
avec = c("PAX6", "ZIC2", "OCT4", "SOX2")
avec
```
16. Subset the vector using [] notation, and get the 5th and 6th elements. 
```{r }
avec[c(5,6)]
```
17. You can also subset any vector using a logical vector in []. Run the following:
```{r }
myvec=1:5
# the length of the logical vector
# should be equal to length(myvec)
myvec[c(TRUE,TRUE,FALSE,FALSE,FALSE)]
myvec[c(TRUE,FALSE,FALSE,FALSE,TRUE)]
```
18. ==,>,<, >=, <= operators create logical vectors. See the results of the following operations:
```{r }
myvec > 3
myvec == 4
myvec <= 2
myvec != 4
```

19. Use the > operator in myvec[ ] to get elements larger than 2 in myvec which is described above.
```{r }
myvec[myvec > 2]
```
20. Make a 5x3 matrix (5 rows, 3 columns) using matrix().
```{r }
mat = matrix(1:15, nrow = 5, ncol = 3)
mat
```
21. What happens when you use byrow = TRUE in your matrix() as an additional argument?
```{r }
mat = matrix(1:15, nrow = 5, ncol = 3, byrow = TRUE)
mat
```
22. Extract the first 3 columns and first 3 rows of your matrix using [] notation.
```{r }
mat[1:3,1:3]
```
23. Extract the last two rows of the matrix you created earlier.
```{r }
mat[4:5,]
```
24. Extract the first two columns and run class() on the result.
```{r }
class(mat[,1:2])
```
25. Extract the first column and run class() on the result, compare with the above exercise.
```{r }
class(mat[,1])
```
26. Make a data frame with 3 columns and 5 rows. Make sure first column is a sequence of numbers 1:5, and second column is a character vector.
```{r }
df=data.frame(col1=1:5,col2=c("a","b","c","S","C"),col3=46:50)
df
```
27. Extract the first two columns and first two rows. HINT: Use the same notation as matrices.
```{r }
df[1:2,1:2]
```
28. Extract the last two rows of the data frame you made. HINT: Same notation as matrices.
```{r }
df[4:5,]
```
29. Extract the last two columns using the column names of the data frame you made.
```{r }
df[,c("col2", "col3")]
```
30. Extract the second column using the column names. You can use [] or $ as in lists; use both in two different answers.
```{r }
df$col2
```
```{r }

df[,"col2"]
```
31. Extract rows where the 1st column is larger than 3. 
```{r }
df[df$col1>3,]
```
32. Extract rows where the 1st column is larger than or equal to 3.
```{r }
df[df$col1>=3,]
```
33. Convert a data frame to the matrix. HINT: Use as.matrix(). Observe what happens to numeric values in the data frame.
```{r }
as.matrix(df)
```
34. Make a list using the list() function. Your list should have 4 elements; the one below has 2.
```{r }
mylist = list(a=c(1,2,3),
             b=c("apple","orange"),
             c=c("Maccabi","Haifa"),
             d=c("Champs", "2020-2021"))
mylist
```
35. Select the 1st element of the list you made using $ notation.
```{r }
mylist$a
```
36. Select the 4th element of the list you made earlier using $ notation.
```{r }
mylist$d
```
37. Select the 1st element of your list using [ ] notation.
```{r }
mylist["a"]
```
38. Select the 4th element of your list using [ ] notation.
```{r }
mylist["d"]
```
39. Make a factor using factor(), with 5 elements.
```{r }
fa = factor(c("1", "2", "3", "4", "5"))
fa
```
40. Convert a character vector to a factor using as.factor(). First, make a character vector using c() then use as.factor().
```{r }
cvec = c("c","h","a","r")
cvec
cvec_fac = as.factor(cvec)
cvec_fac
```
41. Convert the factor you made above to a character using as.character().
```{r }
as.character(cvec_fac)
```


## Reading in and writing data out in R

1. Read CpG island (CpGi) data from the compGenomRData package CpGi.table.hg18.txt. This is a tab-separated file. Store it in a variable called cpgi.
```{r }
cpgFilePath=system.file("extdata", "CpGi.table.hg18.txt", package="compGenomRData")
cpgi=read.table(cpgFilePath, header=TRUE, sep="\t")
```
2. Use head() on CpGi to see the first few rows. 
```{r }
head(cpgi)
```
3. Why doesn’t the following work? See sep argument at help(read.table). \
*Doesn't work because the delimiter is tab and not a comma.*
```{r }
cpgiSepComma=read.table(cpgFilePath, header=TRUE,sep=",")
head(cpgiSepComma)
```
4. What happens when you set stringsAsFactors=FALSE in read.table()? \
*Strings, like the chromosomes' names, won't be converted to factors.*
```{r }
cpgiHF=read.table(cpgFilePath, header=TRUE, sep="\t", stringsAsFactors=FALSE)
head(cpgiHF)
```
5. Read only the first 10 rows of the CpGi table. 
```{r }
cpgi_10rows=read.table(cpgFilePath, header=TRUE, nrows = 10, sep="\t")
dim(cpgi_10rows)
head(cpgi_10rows)
```
6. Use cpgFilePath=system.file("extdata","CpGi.table.hg18.txt", package="compGenomRData") to get the file path, then use read.table() with argument header=FALSE. Use head() to see the results. \
*Headers are now just a regular row.*
```{r }
cpgi_no_header=read.table(cpgFilePath, header=FALSE, sep="\t")
head(cpgi_no_header)
```
7. Write CpG islands to a text file called “my.cpgi.file.txt”. Write the file to your home folder.
```{r }
write.table(cpgi, "my.cpgi.file.txt")
write.table(cpgi, "my.cpgi.file2.txt", quote=FALSE,sep="\t")
```
9. Write out the first 10 rows of the cpgi data frame. HINT: Use subsetting for data frames we learned before.
```{r }
write.table(cpgi[1:10,], "my.cpgi.file_10rows.txt", quote=FALSE,sep="\t")
```
10. Write the first 3 columns of the cpgi data frame.
```{r }
write.table(cpgi[,1:3], "my.cpgi.file_3cols.txt", quote=FALSE,sep="\t")
```
11. Write CpG islands only on chr1. HINT: Use subsetting with [], feed a logical vector using == operator.
```{r }
write.table(cpgi[cpgi$chrom == "chr1",], "my.cpgi.file_chr1.txt", quote=FALSE,sep="\t")
```
12. Read two other data sets “rn4.refseq.bed” and “rn4.refseq2name.txt” with header=FALSE, and assign them to df1 and df2 respectively. They are again included in the compGenomRData package, and you can use the system.file() function to get the file paths.
```{r }
rn4File1Path=system.file("extdata", "rn4.refseq.bed", package="compGenomRData")
rn4File2Path=system.file("extdata", "rn4.refseq.bed", package="compGenomRData")
df1 = read.table(rn4File1Path, header=FALSE)
df2 = read.table(rn4File2Path, header=FALSE)
```
13. Use head() to see what is inside the data frames above.
```{r }
head(df1)
```
```{r }
head(df2)
```

14. Merge data sets using merge() and assign the results to a variable named ‘new.df’, and use head() to see the results.
```{r }
new.df = merge(df1,df2)
head(new.df)
```

## Plotting in R

Please run the following code snippet for the rest of the exercises.
```{r }
set.seed(1001)
x1=1:100+rnorm(100,mean=0,sd=15)
y1=1:100
```

1. Make a scatter plot using the x1 and y1 vectors generated above.
2. Use the main argument to give a title to plot() as in plot(x,y,main="title").
3. Use the xlab argument to set a label for the x-axis. Use ylab argument to set a label for the y-axis.
4. Once you have the plot, run the following expression in R console. mtext(side=3,text="hi there") does. 
```{r }
plot(x1,y1, main = "Scatter plot", xlab = "X axis", ylab = "Y axis")
mtext(side=3,text="hi there")
``` 

5. See what mtext(side=2,text="hi there") does. Check your plot after execution. 
```{r }
plot(x1,y1, main = "Scatter plot", xlab = "X axis", ylab = "Y axis")
mtext(side=2,text="hello there")
```

6. Use mtext() and paste() to put a margin text on the plot. You can use paste() as ‘text’ argument in mtext(). HINT: mtext(side=3,text=paste(...)). See how paste() is used for below. 
```{r }
plot(x1,y1, main = "Scatter plot", xlab = "X axis", ylab = "Y axis")
mtext(side=2,text=paste("General", "Kenobi"))
```

7. cor() calculates the correlation between two vectors. Pearson correlation is a measure of the linear
correlation (dependence) between two variables X and Y. Try using the cor() function on the x1 and
y1 variables.
```{r }
cor(x1,y1)
```
8. Try to use mtext(),cor() and paste() to display the correlation coefficient on your scatter plot.
```{r }
plot(x1,y1, main = "Scatter plot", xlab = "X axis", ylab = "Y axis")
mtext(side=3,text=paste("Correlation:", cor(x1,y1)))
```

9. Change the colors of your plot using the col argument. Ex: plot(x,y,col="red").
```{r }
plot(x1,y1, main = "Scatter plot", xlab = "X axis", ylab = "Y axis", col = "red")
```

10. Use pch=19 as an argument in your plot() command.
```{r }
plot(x1,y1, main = "Scatter plot", xlab = "X axis", ylab = "Y axis", pch = 19)
```

11. Use pch=18 as an argument to your plot() command.
```{r }
plot(x1,y1, main = "Scatter plot", xlab = "X axis", ylab = "Y axis", pch = 18)

```

12. Make a histogram of x1 with the hist() function. A histogram is a graphical representation of the data distribution.
```{r }
hist(x1)
```

13. You can change colors with ‘col’, add labels with ‘xlab’, ‘ylab’, and add a ‘title’ with ‘main’ arguments. Try all these in a histogram.
```{r }
hist(x1, main = "x1 histogram", col = "green3", xlab = "X axis", ylab = "Frequency")

```

14. Make a boxplot of y1 with boxplot().
```{r }
boxplot(y1)
```

15. Make boxplots of x1 and y1 vectors in the same plot.
```{r }
boxplot(x1, y1)

```

16. In boxplot, use the horizontal = TRUE argument.
```{r }
boxplot(x1, y1, horizontal = TRUE)
```

17. Make multiple plots with par(mfrow=c(2,1))
• run par(mfrow=c(2,1))
• make a boxplot
• make a histogram
```{r }
par(mfrow=c(2,1))
boxplot(x1, y1, horizontal = TRUE)
hist(x1, main = "x1 histogram", col = "green3", xlab = "X axis", ylab = "Frequency")
```

18. Do the same as above but this time with par(mfrow=c(1,2)).
```{r }
par(mfrow=c(1,2))
boxplot(x1, y1, horizontal = TRUE)
hist(x1, main = "x1 histogram", col = "green3", xlab = "X axis", ylab = "Frequency")
```

19. Save your plot using the “Export” button in Rstudio.
20. You can make a scatter plot showing the density of points rather than points themselves. If you use
points it looks like this
```{r }
x2=1:1000+rnorm(1000,mean=0,sd=200)
y2=1:1000
plot(x2,y2,pch=19,col="blue")
```

If you use the smoothScatter() function, you get the densities.
```{r }
smoothScatter(x2,y2,
colramp=colorRampPalette(c("white","blue",
"green","yellow","red")))
```

Now, plot with the colramp=heat.colors argument and then use a custom color scale using the following
argument.
```{r }
smoothScatter(x2,y2, colramp = heat.colors)
```

and then use a custom color scale
```{r }
smoothScatter(x2,y2, colramp = colorRampPalette(c("white","green4", "yellowgreen","green2","lightblue")))
```

## Functions and control structures (for, if/else, etc.)

Read CpG island data as shown below for the rest of the exercises.
```{r }
cpgtFilePath=system.file("extdata", "CpGi.table.hg18.txt", package="compGenomRData")
cpgi=read.table(cpgtFilePath,header=TRUE,sep="\t")
head(cpgi)
```

1. Check values in the perGc column using a histogram. The ‘perGc’ column in the data stands for GC percent => percentage of C+G nucleotides. 
```{r }
hist(cpgi$perCpg, main = "%GC in cpgi dataset", xlab = "%GC" )
```

2. Make a boxplot for the ‘perGc’ column.
```{r }
boxplot(cpgi$perCpg, main = "%GC in cpgi dataset", ylab = "%GC" )
```

3. Use if/else structure to decide if the given GC percent is high, low or medium. If it is low, high, or medium: low < 60, high>75, medium is between 60 and 75; use greater or less than operators, < or >.  
Fill in the values in the code below, where it is written ‘YOU_FILL_IN’

```{r }
GCper=65
# check if GC value is lower than 60,
# assign "low" to result
if(GCper < 60){
        result="low"
        cat("low")
} else if(GCper > 75) { # check if GC value is higher than 75,
        #assign "high" to result
        result="high"
        cat("high")
} else{ # if those two conditions fail then it must be "medium"
        result="medium"
}
result
```

4. Write a function that takes a value of GC percent and decides if it is low, high, or medium: low < 60, high>75, medium is between 60 and 75. Fill in the values in the code below, where it is written
‘YOU_FILL_IN’. 

```{r }
GCclass<-function(my.gc){
if(my.gc < 60){
        result="low"
} else if(my.gc > 75) { # check if GC value is higher than 75,
        #assign "high" to result
        result="high"
} else{ # if those two conditions fail then it must be "medium"
        result="medium"
}
return(result)
}
GCclass(10) # should return "low"
GCclass(90) # should return "high"
GCclass(65) # should return "medium"
```

5. Use a for loop to get GC percentage classes for gcValues below. Use the function you wrote above.
```{r }
gcValues=c(10,50,70,65,90)
for( i in gcValues){
        print(GCclass(i))
}
```

6. Use lapply to get GC percentage classes for gcValues. 
```{r }
lapply(gcValues, GCclass)
```

7. Use sapply to get values to get GC percentage classes for gcValues.
```{r }
sapply(gcValues, GCclass)
```

8. Is there a way to decide on the GC percentage class of a given vector of GCpercentages without using
if/else structure and loops ? if so, how can you do it? HINT: Subsetting using < and > operators.

```{r }
set.seed(42)
gcValues = data.frame(perCpg = rnorm(100, 65, 10))
gcValues$perCpg_class = "medium"

gcValues$perCpg_class[gcValues$perCpg > 75] = "high"
gcValues$perCpg_class[gcValues$perCpg < 60] = "low"

gcValues$perCpg_class = factor(gcValues$perCpg_class, levels = c("low", "medium", "high"))

head(gcValues, n = 10)
summary(gcValues$perCpg_class)
        
```



