---
title: "Ex3"
author:
        - "Shaqed Carasso" 
        - "ID: 201281805"
date: "13/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(pheatmap)
library(patchwork)
library(grid)
library(gridExtra)
library(stats)
library(corrplot)
library(ggplot2)
library(ggfortify)
library(ggpubr)
library(DESeq2)
library(edgeR)
library(compcodeR)
library(rpanel)


```

## Differential expression analysis
• Import the read counts and colData tables.

```{r }
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
package = "compGenomRData")

counts <- read.table(counts_file, header = T, sep = '\t')
counts <- counts %>% select(-width)
metadata <-  read.table(coldata_file, header = T, sep = '\t', 
                      stringsAsFactors = TRUE)
```

• Set up a DESeqDataSet object.

```{r }
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design= ~group)

```

• Filter out genes with low counts.

```{r }
threshold <- rowSums(DESeq2::counts(dds)) >= 10

dds <- dds[threshold, ]


```

• Run DESeq2 contrasting the CASE sample with CONTROL samples.

```{r }

dds <- DESeq(dds)

res <- results(dds, contrast = c("group", 'CASE', 'CTRL'))
res_df <-as.data.frame(res)
dds_sig <- res_df %>% mutate(threshold = (padj < 0.05 & abs(log2FoldChange) >= 2)) %>%
        filter(threshold == TRUE)


dds_sig_up <- dds_sig %>% filter(log2FoldChange > 0)
dds_sig_down <- dds_sig %>% filter(log2FoldChange < 0)



length(rownames(dds_sig_up))
length(rownames(dds_sig_down))

```


1. Make a volcano plot using the differential expression analysis results. (Hint: x-axis denotes the
log2FoldChange and the y-axis represents the -log10(pvalue)).

```{r }
# Create the volcano plot

ggplot()+
        geom_point(data = res_df, 
                   mapping = aes(x = log2FoldChange, y = -log10(pvalue)), 
                   col = "grey80", size =1) +
        geom_point(data = subset(res_df, log2FoldChange > 2 & padj < 0.05), 
                   aes(log2FoldChange, -log10(pvalue)), color = "red", size =1)+
        geom_point(data = subset(res_df, log2FoldChange < -2 & padj < 0.05), 
                   aes(log2FoldChange, -log10(pvalue)), color = "steelblue2", size =1)+
        theme_bw()+
        theme(legend.title =element_blank())+
        labs(x = "log2 (Fold change of case vs. control)", y= "-log10(P-value)")

```

2. Use DESeq2::plotDispEsts to make a dispersion plot and find out the meaning of this plot. 

```{r }
plotDispEsts(dds)
```

3. Explore lfcThreshold argument of the DESeq2::results function. What is its default value? What
does it mean to change the default value to, for instance, 1?

```{}
results(object, contrast, name, lfcThreshold = 0,
  altHypothesis = c("greaterAbs", "lessAbs", "greater", "less"),
  listValues = c(1, -1), cooksCutoff, independentFiltering = TRUE,
  alpha = 0.1, filter, theta, pAdjustMethod = "BH", filterFun,
  format = c("DataFrame", "GRanges", "GRangesList"), test,
  addMLE = FALSE, tidy = FALSE, parallel = FALSE,
  BPPARAM = bpparam(), minmu = 0.5)
```

*As we can see in the function's arguments, the default value of lfcThreshold is 0.*
*The argument specify a threshold of minimum absolute log2 fold change for a gene to be considered differentialy expressed.*
*Changing its value to 1, will mean that only genes with log2 fold change of more than 1 or less than -1 between the conditions and with FDR of less than specified alpha will be considered significantly differentialy expressed.*

4. What is independent filtering? What happens if we don’t use it? Google independent filtering
statquest and watch the online video about independent filtering.

*Filtering out genes with low read counts that might affect the results by producing type I error - false positives.*
*If we don't use it, we might falsely identify genes as significantly changed between conditions although they are not.*


5. Re-do the differential expression analysis using the edgeR package. Find out how much DESeq2 and
edgeR agree on the list of differentially expressed genes. 


```{r }
#Create counts object
dgList <- DGEList(counts=counts, genes=rownames(counts))

#filter the same genes that we filtered for DESeq2.
dgList <- dgList[threshold, ]

#set up the experimental design

design <- model.matrix(~metadata$group)
rownames(design) <- rownames(metadata)

#normalize and estimate dispersion

dgList <- calcNormFactors(dgList)
dgList <- estimateDisp(dgList, design,robust=TRUE,prior.df=1)

#run EdgeR
fit <- glmQLFit(dgList, design)
lrt <- glmQLFTest(fit)
dge<-as.data.frame(topTags(lrt,n=Inf))
dge$dispersion<-lrt$dispersion

# extract significant genes
dge<-merge(dge,lrt$fitted.values,by='row.names')
rownames(dge)=dge$Row.names
dge$Row.names=NULL
dge<-dge[order(dge$PValue),]
dge_sig <- dge %>% mutate(threshold = (FDR < 0.05 & abs(logFC) >= 2)) %>%
        filter(threshold == TRUE)


dge_sig_up <- dge_sig %>% filter(logFC > 0)
dge_sig_down <- dge_sig %>% filter(logFC < 0)



length(rownames(dge_sig_up))
length(rownames(dge_sig_down))

```
```{r echo=FALSE}
# check how much they agree

mean_agree <- round(mean(rownames(dge_sig) %in% rownames(dds_sig))*100,2)

print(paste(as.character(mean_agree), "% ", "of the genes in that were identified by edgeR are also in DESeq2 results.", sep = ""))

missing <- sum(!(rownames(dge_sig) %in% rownames(dds_sig)))
missing2 <- sum(!(rownames(dds_sig) %in% rownames(dge_sig)))


print(paste(missing, " Genes in that were identified by edgeR are missing from DESeq2 results.", sep = ""))
print(paste("and ",missing2, " Genes in that were identified by DESeq2 are missing from edgeR results.", sep = ""))


```
```{r }
ggplot()+
  geom_point(data = dge, 
             mapping = aes(x = logFC, y = -log10(PValue)), 
             col = "grey80", size =1) +
  geom_point(data = subset(dge, logFC > 2 & FDR < 0.05), 
             aes(logFC, -log10(PValue)), color = "red", size =1)+
  geom_point(data = subset(dge, logFC < -2 & FDR < 0.05), 
             aes(logFC, -log10(PValue)), color = "steelblue2", size =1)+
  theme_bw()+
  theme(legend.title =element_blank())+
  labs(x = "log2 (Fold change of case vs. control)", y= "-log10(P-value)")

```

6. Use the compcodeR package to run the differential expression analysis using at least three different tools and compare and contrast the results following the compcodeR vignette.
```{r }

count.matrix <- counts
sample.annot <- metadata
colnames(sample.annot)[2] <- "condition"


info.parameters <- list(dataset = "mytestdata", uID = "123456")
cpd <- compData(count.matrix = count.matrix, 
                sample.annotations = sample.annot, 
                info.parameters = info.parameters)

saveRDS(cpd, "saveddata.rds")

runDiffExp(data.file = "saveddata.rds",
           result.extent = "voom.limma", Rmdfunction = "voom.limma.createRmd",
           output.directory = ".", norm.method = "TMM")
runDiffExp(data.file = "saveddata.rds",
           result.extent = "edgeR.exact", Rmdfunction = "edgeR.exact.createRmd",
           output.directory = ".", norm.method = "TMM",
           trend.method = "movingave", disp.type = "tagwise")
runDiffExp(data.file = "saveddata.rds", result.extent = "DESeq2",
           Rmdfunction = "DESeq2.createRmd", test = "Wald",
           output.directory = ".", fit.type = "parametric")


runComparisonGUI(input.directories = ".",
                 output.directory = ".", recursive = FALSE)

```