---
title: "Ex4"
author:
        - "Shaqed Carasso" 
        - "ID: 201281805"
date: "22/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(pheatmap)
library(patchwork)
library(grid)
library(gridExtra)
library(stats)
library(corrplot)
library(ggplot2)
library(ggfortify)
library(ggpubr)
library(DESeq2)
library(edgeR)
library(compcodeR)
library(rpanel)
library(compGenomRData)
library(gProfileR)
library(gage)
```

## Functional enrichment analysis

```{r message=FALSE, warning=FALSE, include=FALSE}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
                           package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
                            package = "compGenomRData")
counts <- as.matrix(read.table(counts_file, header = T, sep = '\t'))

#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))

#define the experimental setup
colData <- read.table(coldata_file, header = T, sep = '\t',
                      stringsAsFactors = TRUE)

#define the design formula
designFormula <- "~ group"

#create a DESeq dataset object from the count matrix and the colData
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = as.formula(designFormula))
dds <- DESeq(dds)

DEresults = results(dds, contrast = c("group", 'CASE', 'CTRL'))
```

```{r message=FALSE, warning=FALSE, include=FALSE}
#remove genes with NA values
DE <- DEresults[!is.na(DEresults$padj),]

#select genes with adjusted p-values below 0.1
DE <- DE[DE$padj < 0.1,]

#select genes with absolute log2 fold change above 1 (two-fold change)
DE <- DE[abs(DE$log2FoldChange) > 1,]

#get the list of genes of interest
genesOfInterest <- rownames(DE)

#calculate enriched GO terms
goResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'GO',
                       hier_filtering = 'moderate')
```

1. Re-run gProfileR, this time using pathway annotations such as KEGG, REACTOME, and protein complex databases such as CORUM, in addition to the GO terms. Sort the resulting tables by columns `precision` and/or `recall`. How do the top GO terms change when sorted for `precision`, `recall`, or `p.value`? hint: use `order()` for sorting.

*It changes the top terms, although all of them still pass by p-values of less than 0.05. *
*By looking at the top precision we can see that 20% of the genes in the input list were annotated to Metabolic pathways *
*Looking at recall we can see the top terms in which the input list contains most of the genes of those terms. *
*By p-value, we look at the top terms that were identified in the input, after correction. *


```{r message=FALSE, warning=FALSE}
KEGGResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'KEGG',
                       hier_filtering = 'moderate')
```

sorted by precision:
The proportion of genes in the input list that are annotated to the function.

```{r message=FALSE, warning=FALSE}

KEGGResults %>% 
        arrange(desc(precision)) %>%
        select(term.name, p.value, precision, recall) %>%
        head()

```
sorted by recall:
The proportion of functionally annotated genes that the query recovers

```{r message=FALSE, warning=FALSE}

KEGGResults %>% 
        arrange(desc(recall)) %>%
        select(term.name, p.value, precision, recall) %>%
        head()

```
```{r message=FALSE, warning=FALSE}

KEGGResults %>% 
        arrange(p.value) %>%
        select(term.name, p.value, precision, recall) %>%
        head()

```

## Gene set enrichment analysis

We use the bioconductor package gage to demonstrate how to do GSEA using normalized expression data of the samples as input.

```{r}
#Let's define the first gene set as the list of genes from one of the
#significant GO terms found in the GO analysis. order go results by pvalue
goResults <- goResults[order(goResults$p.value),]

#restrict the terms that have at most 100 genes overlapping with the query
go <- goResults[goResults$overlap.size < 100,]

geneSet1 <- unlist(strsplit(go[1,]$intersection, ','))

#Define another gene set by just randomly selecting 25 genes from the counts
#table get normalized counts from DESeq2 results
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)

geneSet2 <- sample(rownames(normalizedCounts), 25)

geneSets <- list('top_GO_term' = geneSet1,
                 'random_set' = geneSet2)

#Use the normalized counts to carry out a GSEA.
gseaResults <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'as.group')

gseaResults
```

2. Repeat the gene set enrichment analysis by trying different options for the `compare` argument of the `GAGE:gage`
function. How do the results differ? 

*I've changed it to 'paired', and now I got the results for each case separately rather than a group against group comparison. *

```{r}
gseaResults <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'paired')

gseaResults
```
3. Make a scatter plot of GO term sizes and obtained p-values by setting the `gProfiler::gprofiler` argument `significant = FALSE`. Is there a correlation of term sizes and p-values? (Hint: Take -log10 of p-values). If so, how can this bias be mitigated?

*There seems to be a negative correlation between term size and p-values, meaning that larger term sizes produce more significant results. *
*It would make sense to account in some way the term sizes. maybe by normalizing the identified genes to term size ratio. *

```{r}

goResults_all <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       significant = F,
                       src_filter = 'GO',
                       hier_filtering = 'none'
                       )


ggplot(goResults_all, aes(x = term.size, y = -log10(p.value))) +
        geom_point(aes(color = p.value))+
        scale_colour_gradient(low = "green", high="gray", breaks = c(0.01,0.1,0.2), limits = c(0,0.2))+
        geom_smooth(method = "lm", se = F)+
        theme_pubr(legend = "right")

```


4. Do a gene-set enrichment analysis using gene sets from top 10 GO terms. 

```{r}

goResults <- goResults[order(goResults$p.value),]

go <- goResults[goResults$overlap.size < 100,]

#Define another gene set by just randomly selecting 25 genes from the counts
#table get normalized counts from DESeq2 results
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)

# extracting the terms of top 10 to a list
geneSets10 <- list()

for(i in seq(1,10)) {
    termname = go$term.name[i]
    geneset <- unlist(strsplit(go[i,]$intersection, ','))
    geneSets10[[termname]] <- geneset
}


# Using the defined gene sets, weâ€™d like to do a group comparison between the case
# samples with respect to the control samples.

#Use the normalized counts to carry out a GSEA.
gseaResults <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets10, compare = 'as.group')

gseaResults
```

5. What are the other available R packages that can carry out gene set enrichment analysis for RNA-seq datasets? 

*Packages: clusterProfiler, Enrichr, hypeR, topGO. *

6.  Use the topGO package (https://bioconductor.org/packages/release/bioc/html/topGO.html) to re-do the GO term analysis. Compare and contrast the results with what has been obtained using the `gProfileR` package. Which tool is faster, `gProfileR` or topGO? Why? 

*The results are very different. Only 13 GO terms appeared in both. `gProfileR` worked faster, because it didn't need to download all the GO ENSEMBL database. *

```{r message=FALSE, warning=FALSE}
library(topGO)
library(GO.db)
library(biomaRt)



allgenes <- rownames(DEresults)

#adapted from https://datacatz.wordpress.com/2018/01/19/gene-set-enrichment-analysis-with-topgo-part-1/

# create GO db for genes to be used using biomaRt - please note that this takes a while
db= useMart('ENSEMBL_MART_ENSEMBL',dataset='hsapiens_gene_ensembl', host="www.ensembl.org")
go_ids= getBM(attributes=c('go_id', 'external_gene_name', 'namespace_1003'), filters='external_gene_name', values=allgenes, mart=db)

# build the gene 2 GO annotation list (needed to create topGO object)
gene_2_GO=unstack(go_ids[,c(1,2)])
 
# remove any candidate genes without GO annotation
keep = genesOfInterest %in% go_ids[,2]
keep =which(keep==TRUE)
genesOfInterest=genesOfInterest[keep]
 

# make named factor showing which genes are of interest
geneList=factor(as.integer(allgenes %in% genesOfInterest))
names(geneList)= allgenes

sampleGOdata <- new("topGOdata",
                    description = "Simple session", ontology = "BP",
                    allGenes = geneList, 
                    #geneSel = t(genesOfInterest),
                    nodeSize = 10,
                    annot = annFUN.gene2GO, gene2GO = gene_2_GO)


# define test using the weight01 algorithm (default) with fisher
weight_fisher_result=runTest(sampleGOdata, algorithm='weight01', statistic='fisher') 
 
# generate a table of results: we can use the GenTable function to generate a summary table with the results from tests applied to the topGOdata object.
allGO=usedGO(sampleGOdata)
all_res=GenTable(sampleGOdata, weightFisher=weight_fisher_result, orderBy='weightFisher', topNodes=length(allGO))

#performing BH correction on our p values
p.adj=round(p.adjust(all_res$weightFisher,method="BH"),digits = 4)
 
# create the file with all the statistics from GO analysis
all_res_final=cbind(all_res,p.adj)
all_res_final=all_res_final[order(all_res_final$p.adj),]
 
#get list of significant GO before multiple testing correction
results.table.p= all_res_final[which(all_res_final$weightFisher<=0.001),]
 
#get list of significant GO after multiple testing correction
results.table.bh=all_res_final[which(all_res_final$p.adj<=0.05),]

#number of genes that appear in both
sum( goResults$term.name %in% results.table.bh$Term)

head(results.table.bh$Term, n = 10)
head(goResults$term.name, n = 10)

```
7. Given a gene set annotated for human, how can it be utilized to work on _C. elegans_ data? (Hint: See `biomaRt::getLDS`). 

*Change the dataset argument in the useMart function to 'celegans_gene_ensembl'*


8. Import curated pathway gene sets with Entrez identifiers from the [MSIGDB database](http://software.broadinstitute.org/gsea/msigdb/collections.jsp) and re-do the GSEA for all curated gene sets. 

```{r }

library(msigdbr)

cgp_gene_sets = msigdbr(species = "Homo sapiens", category = "C2")

msigdbr_list = split(x = cgp_gene_sets$gene_symbol, f = cgp_gene_sets$gs_name)

gseaResults_cgp <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = msigdbr_list, compare = 'as.group')


gseaResults
```