---
title: "Final Project - Genomic Data Science Course, Winter 2022"
author: 
        - "Shaqed Carasso" 
        - "ID: 201281805"
date: "3/13/2022"
output: html_document
---

<style>
p.comment {
background-color: #DBDBDB;
padding: 10px;
border: 1px solid black;
margin-left: 0px;
border-radius: 5px;
}

</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Deletion of Trim28 in committed adipocytes promotes obesity but preserves glucose tolerance

This work is presented as a final project for the course Genomic Data Science (138046), the Technion, Israel Institute of Technology.

In this project, I present a bulk RNAseq analysis - using the published sequencing results available from: 

<p class="comment">
Deletion of Trim28 in committed adipocytes promotes obesity but preserves glucose tolerance. Bond, S.T., King, E.J., Henstridge, D.C. et al.  
_Nat Commun_ 12, 74 (2021). https://doi.org/10.1038/s41467-020-20434-3
</p>

In the aforementioned paper, the researchers focus on a gene proposed to primarily influence adiposity via epigenetic mechanisms in embryonic development (tripartite motif containing 28 - Trim28).

The authors show that mutant mice lacking the Trim28 gene in committed adipocytes develop obesity similar to global Trim28 deletion models. 
This results show that Trim28 has a post-developmental role - affecting gene expression patterns of the adipose tissues.
The effects of the knock-out mutation were xacerbated in female mice, demonstrating a sex-specific regulation of the Trim28 gene.

I will look into the differences between the Trim28-KO and WT mice, and also between male and females Trim-KO mice - elucidating the general and sex specific effects of the gene.

#### Loading neccessary libraries

```{r message=FALSE, warning=FALSE, include=TRUE, paged.print=FALSE}

library(ggplot2)
library(MetBrewer)
library(DESeq2)
library(pheatmap)
library(tidyverse)
library(ggpubr)
library(limma)
library(reshape2)
library(RColorBrewer)
library(corrplot)
library(patchwork)
library(dplyr)

```

#### Reading the dataset 

I'm using the `read.table` function to open the sequencing results (counts) and the experimental design (metadata).

```{r open files}

counts_file <- "trim28_project/E-MTAB-9809-raw-counts.tsv"
coldata_file <- "trim28_project/E-MTAB-9809-experiment-design.tsv"

counts <- read.table(counts_file, sep = '\t',header = T, row.names = 1, stringsAsFactors = TRUE)
metadata <-  read.table(coldata_file, row.names = 1, header = T, sep = '\t', 
                        stringsAsFactors = TRUE)
```

Creating a 'dictionary' of gene ids and names, then remove gene names from counts file

```{r gene names}
geneNames <- as.data.frame(counts[, "Gene.Name"])
geneNames$Gene.ID <- rownames(counts)
geneNames <- geneNames %>% rename("Gene.Name"='counts[, "Gene.Name"]')
counts <- subset( counts, select = -Gene.Name)
```

Filtering the experimental design file to include only informative columns of metadata, specifically sex and group(Trim-KO or WT)

```{r prepare metadata}
metadata <- metadata %>% select('Sample.Characteristic.genotype.', 'Sample.Characteristic.sex.') %>%
        rename( "Sex"= 'Sample.Characteristic.sex.') %>%
        mutate(Group = ifelse(grepl("Trim28", Sample.Characteristic.genotype.), "Trim28KO", "WT"))


```

#### Visualizing library sizes

Bar plot of library sizes. This visualization serves as a quality control measure to look for libraries with low counts.
We can see that all libraries have more than 10 milion counts.

```{r lib sizes}
metadata$lib_size = colSums(counts)

ggplot(metadata, aes(fill = Group)) +
        geom_col(aes(x = rownames(metadata), y = lib_size)) +
        labs(title = "Barplot of library sizes") +
        #scale_x_discrete(limits = rownames(metadata)) +
        theme_pubr() +
        xlab(NULL) +
        ylab("Library size") +
        scale_fill_manual(values=met.brewer("Isfahan1", 2)) +
        facet_wrap(~Sex, scales = "free") +
        theme(axis.text.x = element_text(angle=90, vjust=0.4))

```

#### Normalization to Counts Per Milion - to filter lowly expressed genes

Normalize to CPM, excluding lowly expressed genes that have less than 3 CPM in at least 3 samples.
Just to note, the authors wrote in the paper's method section that they have filtered for genes with average read counts lower than 30 CPM reads. I think that their cut-off is too strict, and leaves out important genes (including Klf14, which is one of the main findings of the paper, thus I've decided to follow with the threshold of 3 CPM in at least 3 samples.

```{r normalize to cpm}
cpm <- apply(counts, 2, 
             function(x) as.numeric(x)/sum(as.numeric(x)) * 10^6)

thresh <- cpm >= 3
keep <- rowSums(thresh) >= 3

counts.filtered = counts[keep,]
dim(counts.filtered)

```

#### Correlation plot

As another measure of quality control, I'm using the CPM normalized counts to look at the correlations between the samples.
Sample ERR4873315 is very different from the rest and thus will be filtered out of the analysis.


```{r correlation}
# Compute the correlation values between samples
correlationMatrix <- cor(cpm) 

breaksList <- seq(0.5, 1, by = 0.001)

heatmap.corr <- pheatmap(correlationMatrix, annotation = select(metadata, c(Group, Sex)), 
                         annotation_colors = list(Group = c('WT' = "#178f92", Trim28KO = "#845d29"), 
                                                  Sex = c('male' = "lightgreen", 'female' = "salmon" )),
                         breaks = breaksList,
                         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList)),
                         main = "Correlations Heatmap")

```

#### Creating a DESeq2 object for RNAseq differential expression analysis

I combine the metadata and counts files into a DESeq2 object.
The experimental design takes into consideration sex as a confounding factor.


```{r initiate deseq2, message=FALSE, warning=FALSE}
metadata_filtered_samples <- subset(metadata, row.names(metadata) != ("ERR4873315"))

subset_Samples <- as.vector(rownames(metadata_filtered_samples))
subset_reads <- counts.filtered[, subset_Samples]

countdata.deseq <- DESeq2::DESeqDataSetFromMatrix(countData = subset_reads, 
                                                  colData = metadata_filtered_samples, 
                                                  design = ~ Group + Sex)

dds <- DESeq(countdata.deseq)

# Extract the normalized counts
countdata_normalized_counts <- counts(dds, normalized = TRUE)
```
Visualizing the dispersion of the data. 

```{r dispersion}
plotDispEsts(dds)
```

#### Visualizing the distances between samples, using Principle Component Analysis for dimension reduction. 

Samples with similar pattern of expression are closer to each other on the graph.
I have Extracted the transformed values (using Variance stabilizing transformation) and selected the top 1000 most variable genes.

The PCA plot shows that the samples are divided by sex, but also to some extent by their group.


```{r PCA}
# Extracting transformed values, selecting top 1000 most variable genes

vsd <- vst(dds, blind=FALSE, nsub = 1000)

pcaData <- plotPCA(vsd, intgroup = c("Group", "Sex"), returnData = TRUE)

percentVar <- round(100 * attr(pcaData, "percentVar"),digits = 1)

ggplot(pcaData, aes(x = PC1, y = PC2, color = Group)) +
        coord_fixed()+
        geom_point(size = 2, aes(shape = Sex)) +
        xlab(paste0("PC1: ", percentVar[1], "%")) + 
        ylab(paste0("PC2: ", percentVar[2], "%")) +
        labs(color = "Group") +
        guides(color = guide_legend(order = 1),
               shape = guide_legend(order = 2))+
        scale_fill_manual(values=met.brewer("Isfahan1", 2), guide = NULL) +
        scale_color_manual(values=met.brewer("Isfahan1", 2)) +
        theme_pubr(legend = "top")
                
```


#### Differential expression analysis

First, we will compare between the WT and Trim-28 KO mice.
I'm running the DESeq2 package's results function to extract the significantly differentially expressed genes(p-value < 0.05 after FDR (BH)).
I'm not specifying a log fold change threshold, I will filter the results based on adjusted p values and Fold changes,


```{r summarizing results}

res_Group <- results(dds, 
               contrast = c("Group", "Trim28KO", 'WT'), 
               alpha = 0.05, 
               lfcThreshold = 0)

summary(res_Group)
```

We will visualize the results using a volcano plot, coloring only genes that passed the threshold of p.value lower than 0.01, after FDR correction.

```{r volacno, message=FALSE, warning=FALSE}

dds_res_all <- data.frame(res_Group) %>% 
        mutate(threshold = ifelse((padj < 0.01 & log2FoldChange < 0), "Downregulated", 
                                  ifelse((padj < 0.01 & log2FoldChange > 0), "Upregulated", "NS"))) %>%
        filter(!is.na(padj))


dds_res_all <- left_join(rownames_to_column(dds_res_all), geneNames, by=c("rowname" = "Gene.ID"))

library(ggbreak)

ggplot(dds_res_all) + 
        geom_point(aes(x = log2FoldChange, y = -log10(padj), color = threshold)) + 
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") + 
        theme(legend.position = "none", 
              plot.title = element_text(size = rel(1.5), hjust = 0.5), 
              axis.title = element_text(size = rel(1.25)))+
        scale_color_manual(values=c("#178f92","gray", "#845d29")) +
        geom_text(data = subset(dds_res_all, abs(log2FoldChange) > 4 & padj < 0.01),
                  aes(log2FoldChange, -log10(padj), label = Gene.Name), 
                  vjust =1, size = 4)+
        
        geom_text(data = subset(dds_res_all, Gene.Name %in% c("Elovl3", "Olr1", "Elovl7", "Klf14")),
                  aes(log2FoldChange, -log10(padj), label = Gene.Name), 
                  vjust =1, size = 4)+
        scale_x_continuous(limits = c(-7,8))+
        scale_y_continuous(limits = c(-1,130))+
        scale_y_break(c(50,110))+
        geom_hline(yintercept = -log10(0.01), linetype = 2)+
        theme_pubr(legend = "right")


```


As shown in the volcano plot, we can see that the results from the paper are reproduced, showing upregulation of Elovl3, Elovl7 and Olr1, and downregulation of KLf14.

Interstingly, in the paper, the significantly upregulated Tcfl5 gene is not shown. Comparing my results with the published data at the 'Expression Atlas' we can see that it is indeed upregulated in both males and female mice. 

(https://www.ebi.ac.uk/gxa/experiments/E-MTAB-9809/Results?specific=true&geneQuery=%255B%257B%2522value%2522%253A%2522Tcfl5%2522%252C%2522category%2522%253A%2522symbol%2522%257D%255D&filterFactors=%257B%257D&cutoff=%257B%2522foldChange%2522%253A1%252C%2522pValue%2522%253A0.05%257D&regulation=%2522UP_DOWN%2522)

Tcfl5 or 'Transcription Factor Like 5' protein, is a transcription factor that can affect gene expression in the cells.

Using the tool Harmonize, Tcfl15 is confirmed to be targeted by Trim28.

<p class="comment">
Rouillard AD, Gundersen GW, Fernandez NF, Wang Z, Monteiro CD, McDermott MG, Ma'ayan A. The harmonizome: a collection of processed datasets gathered to serve and mine knowledge about genes and proteins. Database (Oxford). 2016 Jul 3;2016. pii: baw100.
<p>
The genes shown in the paper, and the most significant genes from my volcano plot are visualized below in boxplots, comparing the groups and sex of the mice.

Olr1, has been associated with obesity and other metabolic conditions, and shows upregulation in Trim28-KO mice, but also high variance between the mice.

Looking at Elovl3, we can see the sex differences as it is expressed more in Trim28 KO female mice.

another Sex difference is seen in Klf14, that although it is downregulated in both sexes of Trim28 KO groups, it is almost completely ablated in the  female mice.


```{r boxplots specific genes}

gene_list_volcano <- c("Elovl3", "Elovl7", "Tcfl5", "Klf14", "Gm1821", "Fbx040", "Olr1")
data_boxplots_volcano <- melt(countdata_normalized_counts[geneNames$Gene.ID[geneNames$Gene.Name %in% gene_list_volcano],], value.name = "Counts") %>%
        rename(Gene.ID = Var1 ,Sample = Var2)


data_boxplots_volcano <- left_join(data_boxplots_volcano, rownames_to_column(metadata), by=c("Sample" = "rowname")) %>%
        left_join(geneNames[, c("Gene.ID", "Gene.Name")], by = "Gene.ID") %>%
        mutate(Group_sex = factor(paste(Group, Sex, sep = ", ")))

ggplot(data = data_boxplots_volcano, aes(x=Group_sex, y=Counts, fill = Group, color = Group)) +
               geom_point(position=position_jitter(0.05)) + 
        geom_boxplot(position=position_dodge(1), show.legend = FALSE, alpha = 0.4,
                     outlier.shape = NA, aes(fill = Group))+
        facet_wrap(~Gene.Name, scales = "free") + 
        scale_fill_manual(values=met.brewer("Isfahan1", 2)) +
        scale_color_manual(values=met.brewer("Isfahan1", 2)) +
        theme_pubr(base_size = 8, legend = "top")+
        labs(y = "DESeq normalized counts",
             x = NULL) +
        scale_x_discrete(breaks = c("Trim28KO, male", "Trim28KO, female", "WT, male", "WT, female"),
                          labels= c("Male", "Female", "Male", "Female"))+
        labs(fill = "Group") +
        guides(fill = guide_legend(override.aes = list(size = 5)))

```


A heatmap visualization of the differences between the groups show the different expression patterns between Trim28 KO and WT mice, and between males and females.

```{r heatmap}
# Subset normalized counts to significant genes

dds_res_sig <- subset(dds_res_all, padj < 0.01 & abs(log2FoldChange) >= 0)

sig_norm_counts_dds <- countdata_normalized_counts[dds_res_sig$rowname, ]

# Choose heatmap color palette
heat_colors <- brewer.pal(n = 6, name = "YlOrRd")

# Plot heatmap
heatmap.numClusters <- pheatmap(sig_norm_counts_dds, 
                                annotation_colors = list(Group = c('WT' = "#178f92", Trim28KO = "#845d29"), 
                                                  Sex = c('male' = "lightgreen", 'female' = "salmon" )),
                                cluster_rows = T, 
                                show_rownames = F,
                                show_colnames = F,
                                annotation = select(metadata, c(Group, Sex)), 
                                scale = "row")


```
#### Functional enrichment analysis

The next step will be a functional enrichment analysis of the upregulated and downregulated genes lists.
I'm using the 'gprofiler2' package, as it fast, produces interactive visualizations, and can use multiple databases.

I'm extracting the genes that passed a threshold of 0.01 adjusted p-value, creating lists of genes that were up or down regulated (absolute log2FC of more than 0.5).
Also, I'm using the 'AnnotationDbi' package to translate the gene IDs from ENSEMBLE to ENTREZID, so I could input them into the 'gprofiler2' enrichment functions.


```{r enrichment prepare gene lists, message=FALSE, warning=FALSE}
library("AnnotationDbi")
library(org.Mm.eg.db)

res_Group = res_Group[!is.na(res_Group$padj),]

res_Group$ENTREZID_id = mapIds(org.Mm.eg.db,
                            keys=rownames(res_Group), 
                            column="ENTREZID",
                            keytype="ENSEMBL",
                            multiVals="first")

all_genes <- rownames(res_Group$ENTREZID_id)
# Extract significant results
signif_res_up <- res_Group[res_Group$padj < 0.01 & (res_Group$log2FoldChange) > 0.5 & !is.na(res_Group$padj), ]

signif_res_dn <- res_Group[res_Group$padj < 0.01 & (res_Group$log2FoldChange) < -0.5 & !is.na(res_Group$padj), ]

```

Now we can run an enrichment analysis, I'm looking for enriched terms from the KEGG pathways database, the Gene Ontology Biological Process (BP) and Cellular Component (CC) lists, and the Reactome pathways database.

We can see (in REAC, blue), that Trim28 KO mice were downregulated in genes associated with Insulin-like Growth Factor(IGF) transport and uptake, and also in (BP, orange) genes related to cell motility and and development.

Looking at the enriched terms in the upregulated genes, we see enrichment of genes related to fatty-acid metabolism (REAC, blue). We can also observe an increase in terms (BP, orange) related to fatty acid metabolism process and very long chain fatty acids metabolic process (the Elovl gene family).  


```{r visualize enrichment, message=FALSE, warning=FALSE}

library(gprofiler2)


p <- gost(query = list( "Upregulated" = rownames(signif_res_up), "Downregulated" = rownames(signif_res_dn)), 
          organism = "mmusculus",
          multi_query = TRUE,
          significant = TRUE,
          sources = c("KEGG", "GO:BP", "GO:CC", "REAC")
          )

gostplot(p, capped = TRUE, interactive = TRUE)

```


#### Comparing the WT and Trim-28 KO female mice only

Filtering out the male mice, I will showcase the sexual differences by finding the genes that were altered only in female mice.

Filtering males and creating a new DESeq2 object.

```{r initiate deseq2 females only, message=FALSE, warning=FALSE}
metadata_filtered_samples <- subset(metadata_filtered_samples, Sex != "male")

subset_Samples <- as.vector(rownames(metadata_filtered_samples))
subset_reads <- counts.filtered[, subset_Samples]

countdata.deseq <- DESeq2::DESeqDataSetFromMatrix(countData = subset_reads, 
                                                  colData = metadata_filtered_samples, 
                                                  design = ~ Group)

dds_f <- DESeq(countdata.deseq)

```

Extracting the results of the comparison.

```{r summarizing results F}

res_Females <- results(dds_f, 
               contrast = c("Group", "Trim28KO", 'WT'), 
               alpha = 0.05, 
               lfcThreshold = 0)

summary(res_Females)

```
visualizing the results in a volcano plot.

```{r volacno F}

dds_res_females <- data.frame(res_Females) %>% 
        mutate(threshold = ifelse((padj < 0.01 & log2FoldChange < 0), "Downregulated", 
                                  ifelse((padj < 0.01 & log2FoldChange > 0), "Upregulated", "NS"))) %>%
        filter(!is.na(padj))

dds_res_females <- left_join(rownames_to_column(dds_res_females), geneNames, by=c("rowname" = "Gene.ID"))

ggplot(dds_res_females) + 
        geom_point(aes(x = log2FoldChange, y = -log10(padj), color = threshold)) + 
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") + 
        theme(legend.position = "none", 
              plot.title = element_text(size = rel(1.5), hjust = 0.5), 
              axis.title = element_text(size = rel(1.25)))+
        scale_color_manual(values=c("#178f92","gray", "#845d29")) +
        geom_text(data = subset(dds_res_females, abs(log2FoldChange) > 2 & padj < 0.00000001),
                  aes(log2FoldChange, -log10(padj), label = Gene.Name), 
                  vjust =1, size = 4)+
        
        geom_text(data = subset(dds_res_females, Gene.Name %in% c("Elovl3", "Olr1", "Elovl7", "Klf14")),
                  aes(log2FoldChange, -log10(padj), label = Gene.Name), 
                  vjust =1, size = 4)+
        scale_x_continuous(limits = c(-7,8))+
        geom_hline(yintercept = -log10(0.01), linetype = 2)+
        theme_pubr(legend = "right")


```


We can see that Elovl3 is below the threshold, echoing the differences between male and female mice we saw in the boxplots earlier. 

Now, I will look for genes that were differentially expressed between Trim KO and WT female mice, that were not significantly changed in male mice.
The Gene will then be visualized in boxplots.

```{r boxplots specific genes F}

gene_list_all <- (dds_res_all %>% filter(padj < 0.01 & abs(log2FoldChange) > 2))$Gene.Name

gene_list_females <- (dds_res_females %>% filter(padj < 0.01 & abs(log2FoldChange) > 2))$Gene.Name

gene_list_cut <- gene_list_females[!(gene_list_females %in% gene_list_all)]

data_boxplots_volcano <- melt(countdata_normalized_counts[geneNames$Gene.ID[geneNames$Gene.Name %in% gene_list_cut],], value.name = "Counts")                                    %>% rename(Gene.ID = Var1 ,Sample = Var2)


data_boxplots_volcano <- left_join(data_boxplots_volcano, rownames_to_column(metadata), by=c("Sample" = "rowname")) %>%
        left_join(geneNames[, c("Gene.ID", "Gene.Name")], by = "Gene.ID") %>%
        mutate(Group_sex = factor(paste(Group, Sex, sep = ", ")))

ggplot(data = data_boxplots_volcano, aes(x=Group_sex, y=Counts, fill = Group, color = Group)) +
               geom_point(position=position_jitter(0.05)) + 
        geom_boxplot(position=position_dodge(1), show.legend = FALSE, alpha = 0.4,
                     outlier.shape = NA, aes(fill = Group))+
        facet_wrap(~Gene.Name, scales = "free") + 
        scale_fill_manual(values=met.brewer("Isfahan1", 2)) +
        scale_color_manual(values=met.brewer("Isfahan1", 2)) +
        theme_pubr(base_size = 8, legend = "top")+
        labs(y = "DESeq normalized counts",
             x = NULL) +
        scale_x_discrete(breaks = c("Trim28KO, male", "Trim28KO, female", "WT, male", "WT, female"),
                          labels= c("Male", "Female", "Male", "Female"))+
        labs(fill = "Group") +
        guides(fill = guide_legend(override.aes = list(size = 5)))

```

We can see that most of the genes change in the same direction in both male and females.
Cdkl4 (Cyclin Dependent Kinase Like 4) shows inverted results in males and females. Cdkl4 is upregulated in males, but downregulated in females.

I assume that the genes presented here that changed in female mice didn't pass the significance threshold after adjusting the p-values.

### Applying machine learning to extract features differentiaing between the groups

Although the sample size is (very) low - 24 samples, I will try to use a random forest model to extract features that differentiate between Trim-KO and WT mice - possibly finding more relevant genes.


I'm using the 'Caret' to preform the analysis.

As the sample size is relatively low, I'm splitting the data 60%:40% (13:9 samples), and filtering the data for the 2000 most variable genes.

The model will be based on DESeq2 normalized counts.

I know that the model will be very sensitive to the sampling random seed because of the low sample size and the variance between the samples.
For convenience, and to reproduce my results - 
I will run the analysis with a constant random seed. 



```{r split data, message=FALSE, warning=FALSE}

library(caret)
metadata_filtered_samples <- subset(metadata, row.names(metadata) != ("ERR4873315"))
subset_reads <- counts.filtered[, subset_Samples]

#creating a classification dataframe
class <- DataFrame(Group = factor(metadata_filtered_samples$Group))
rownames(class) <- rownames(metadata_filtered_samples)

#setting seed
set.seed(1201)

#filtering most variable genes
vars <- sort(apply(countdata_normalized_counts, 1, var, na.rm = TRUE), decreasing = TRUE)
data <- countdata_normalized_counts[names(vars)[1:2000], ]

#splitting the data
nTest <- ceiling(ncol(data) * 0.4)
ind <- sample(ncol(data), nTest, FALSE)

# Minimum count is set to 1 in order to prevent 0 division problem within
# classification models.
data.train <- as.matrix(data[ ,-ind] + 1)
data.test <- as.matrix(data[ ,ind] + 1)
classtrain <- DataFrame(condition = class[-ind, ])
classtest <- DataFrame(condition = class[ind, ])


```

create control and train the model. I'm using cross validation to account for the low sample size.

```{r control and train model, message=FALSE, warning=FALSE}

numFolds <- trainControl(method = "repeatedcv", number = 10, repeats = 5)

rf_model <- train(x = t(data.train),
                       y = classtrain$condition,
                       method = "rf",
                       metric = "Accuracy",
                       tuneLength  = 15,
                       trControl = numFolds)

```

Looking at the model, we can see that it was able to predict the classes of the test set correctly. Although it preformed well, showing high sensitivity and specificity, it should be remembered that this results is biased by the initial train-test split. Having imbalance data, confounded by the sex variable - can affect the model's ability to predict correctly, as it might pick up on the differences between the sexes. 

For example, having a majority of males in the train set will overfitting the model to males - leading to worse performance on the test set - having a majority of female samples.

Using 'Caret's tuning - the random forest's best mtry hyperparameter was 23, yielding the best results.

```{r confusion matrix, message=FALSE, warning=FALSE}
show(rf_model)

pred.rf <- predict(rf_model, t(data.test))

pred.rf <- relevel(pred.rf, ref = "WT")
actual <- relevel(classtest$condition, ref = "WT")
tbl <- table(Predicted = pred.rf, Actual = actual)
confusionMatrix(tbl, positive = "WT")

plot(rf_model)
```


Nevertheless, looking at variable importance, we can see that the model was able to pick relevant genes that are different between the classes.

notably, the DJ-1 (PARK7) gene was previously shown to contribute to adipogenesis and obesity-induced inflammation in mice.

<p class="comment">
Kim, JM., Jang, HJ., Choi, S. et al. DJ-1 contributes to adipogenesis and obesity-induced inflammation. _Sci Rep 4_, 4805 (2014). https://doi.org/10.1038/srep04805
</p>

```{r extract important genes}

importance <- varImp(rf_model, scale = TRUE)
importance <- varImp(rf_model)$importance
importance <- left_join(rownames_to_column(importance), geneNames, by=c("rowname" = "Gene.ID"))

library(forcats)

importance %>% arrange(Overall) %>%
        mutate(Gene = fct_reorder(Gene.Name, Overall)) %>%
        top_n(10, Overall) %>%
        ggplot(aes(x = Gene, y = Overall)) +
        geom_segment(aes(xend=Gene), yend=0) +
        labs(title = 'Random Forest (rf package). Gene importance') +
        geom_point(size = 3,shape = 21, colour = "blue", fill = "white") +
        expand_limits(y=0) +
        coord_flip() +
        theme_pubr()

```


```{r plot important genes boxplot}

list_imp_genes <- importance %>% arrange(Overall) %>%
        mutate(Gene = fct_reorder(Gene.Name, Overall)) %>%
        top_n(10, Overall)

data_boxplots <- melt(countdata_normalized_counts[geneNames$Gene.ID[geneNames$Gene.Name %in% list_imp_genes$Gene.Name],], value.name = "Counts") %>%
        rename(Gene.ID = Var1 ,Sample = Var2)


data_boxplots <- left_join(data_boxplots, rownames_to_column(metadata), by=c("Sample" = "rowname")) %>%
        left_join(geneNames[, c("Gene.ID", "Gene.Name")], by = "Gene.ID") %>%
        mutate(Group_sex = factor(paste(Group, Sex, sep = ", ")))


ggplot(data = data_boxplots, aes(x=Group_sex, y=Counts, fill = Group, color = Group)) +
               geom_point(position=position_jitter(0.05)) + 
        geom_boxplot(position=position_dodge(1), show.legend = FALSE, alpha = 0.4,
                     outlier.shape = NA, aes(fill = Group))+
        facet_wrap(~Gene.Name, scales = "free") + 
        scale_fill_manual(values=met.brewer("Isfahan1", 2)) +
        scale_color_manual(values=met.brewer("Isfahan1", 2)) +
        theme_pubr(base_size = 8, legend = "top")+
        labs(y = "DESeq normalized counts",
             x = NULL) +
        scale_x_discrete(breaks = c("Trim28KO, male", "Trim28KO, female", "WT, male", "WT, female"),
                          labels= c("Male", "Female", "Male", "Female"))+
        labs(fill = "Group") +
        guides(fill = guide_legend(override.aes = list(size = 5)))


```

### Conclusions

Overall, the results show different expression patterns in the committed adipocytes of Trim28-KO mice compared to the WT, dependent to some extent on the sex.

DESeq2 analysis revealed sets of genes that were either up or down regulated in KO mice.
Gene ontology enrichment analysis revealed that those genes are related to reduced cell motility and development and increase in fatty acid metabolism - specifically processes related to the metabolism of very long chain fatty acids.

These results suggest that Trim28 can modulate adipose tissue function in a sex-specific manner - contributing to obesity. 

#### results are also available as a shiny app


Github: https://github.com/ShaQasso/GenomicDS_course


ShinyApp: https://shaqed-carasso.shinyapps.io/Final_trim28_project/ 

